### PR TITLE
[SPARK-35396][CORE][FOLLOWUP] Free memory entry immediately

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This is a followup of https://github.com/apache/spark/pull/32534 , and proposes to free the memory entry immediately instead of doing it in the backround asynchronously. The reason is:
1. It's a bit weird to free the resource in an asynchronous way.
2. We free the off-heap memory entry in the same thread, and it's better to be consistent with it.
3. We can simplify the code quite a bit.

This PR also simplifies the tests to reuse the class definition.

### Why are the changes needed?

code simplification

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

existing tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @xuechendi  @srowen @jiangxb1987

**Comment by xuechendi** [comment]
It seems I can’t comment on your PR, from previous discussion, @srowen suggested we use async close to avoid holding memoryStore lock for too long.

**Comment by cloud-fan** [comment]
IIUC, In the real world this feature is for user-defined off-heap-hashRelation for BHJ, and releasing the memory should be fast, at least should be on-par with releasing `SerializedMemoryEntry` by Spark itself. I think it's fine to follow `SerializedMemoryEntry` and release the `AutoCloseable` in the current thread.

If it does introduce perf problems, the Spark vendor that leverages this feature should provide some metrics and propose to release memory entry asynchronously.

**Comment by srowen** [comment]
OK I'm fine with doing in synchronously - I was afraid that closing the objects could be expensive, and, is best-effort in any event, so made sense to do asynchronously.

**Comment by xuechendi** [comment]
> OK I'm fine with doing in synchronously - I was afraid that closing the objects could be expensive, and, is best-effort in any event, so made sense to do asynchronously.

I see, either way is OK to me, and I personally felt non-blocking makes more sense, since these objects will be released eventually, just like other GCable objects stored in MemoryStore, they are also released asynchronously by GC instead of releasing immediately.

**Comment by cloud-fan** [comment]
@xuechendi IIUC your use case is to actually store off-heap data with on-heap memory entry, and I feel it's better to follow the off-heap style and free the memory immediately.

Things may change in the future if we use this feature for real on-heap data, but we can change at that time. e.g. Java may provide new APIs to manage off-heap data and free the memory with GC.

File: core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala [hunk 1]
```diff
@@ -390,39 +387,28 @@ private[spark] class MemoryStore(
...
+  def freeMemoryEntry[T <: MemoryEntry[_]](entry: T): Unit = {
+    entry match {
+      case SerializedMemoryEntry(buffer, _, _) => buffer.dispose()
+      case e: DeserializedMemoryEntry[_] => e.value.foreach {
+        case o: AutoCloseable =>
+          try {
+            o.close()
+          } catch {
+            case NonFatal(e) =>
+              logWarning(s"Got NonFatal exception during remove", e)
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:399]: nit: s"" is not needed.

**Review by viirya — COMMENTED** [review]
Looks okay.

**Review by viirya — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Pull request closed without merging**